### PR TITLE
[#1159] Add spider for farmershomefurniture

### LINKED ISSUE #1159: Farmers home Furniture
http://farmershomefurniture.com/locator

### PR DESCRIPTION
**Pull request by ashikjm**
Fixes #1159 

Sample output: 
`{"type": "Feature", "id": "QjvtM2xnZyU3RTUAr0iFVTx92yo=", "properties": {"ref": "covingtonga", "@spider": "farmershomefurniture", "addr:full": "3197 N ELM ST", "addr:city": "COVINGTON", "addr:state": "GA", "phone": "770-786-1705", "opening_hours": "Mo-Fr 09:00-07:00; Sa 09:00-05:00", "brand": "Farmers Home Furniture"}, "geometry": {"type": "Point", "coordinates": [-83.856689, 33.60239]}},`

### CODE REVIEW

**Comment by iandees** [comment]
Looks like this is failing to fetch because they're giving a 403 response. Can you try changing the user agent?

**Comment by ashikjm** [comment]
Also had to add download_delay to avoid the 429s.

` 'downloader/response_count': 243,
 'downloader/response_status_count/200': 240,
 'downloader/response_status_count/404': 3,`

**Comment by ashikjm** [comment]
@iandees Is there anything that you want me to add in this spider?

**Comment by iandees** [comment]
No, but we need to find a way to make it work before merging.

**Comment by ashikjm** [comment]
I am not seeing any 403s or 429s anymore. Scrapy stats show that out of 243, 240 are 200s and 3 are 404. Am I overlooking something?

` 'downloader/response_count': 243,
 'downloader/response_status_count/200': 240,
 'downloader/response_status_count/404': 3,`

File: locations/spiders/farmershomefurniture.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_1) AppleWebKit/537.36 (KHTML, like Gecko) "
+                  "Chrome/63.0.3239.84 Safari/537.36"
+    }
+
+    def parse(self, response):
+        for store in response.xpath('//tr'):
+            store_data = store.xpath('./td/text()').getall()
+            if store_data:
+                properties={ ['city', 'state', 'address', 'phone'][i]: store_data[i]
+                                                        for i in range(len(store_data)) }
```
Comment (iandees) [inline locations/spiders/farmershomefurniture.py:35]: This for loop is hard to read and seems to be unnecessary. Can you split the four properties into their own lines?

File: locations/spiders/farmershomefurniture.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    def parse_store(self, response, city, state, address, phone):
+        opening_hours = OpeningHours()
+        store_hours = response.xpath('//div[@class="workspacearea"]/div/div/p/text()').extract()[2:]
+
+        for hours in store_hours:
+            day = hours.strip().split(':')[0]
+            if day != 'Sun':
+                time_range= hours.strip().split(':')[1].split('-')
+                if time_range[0] != 'Closed':
+                    opening_hours.add_range(day=DAYS[str(store_hours.index(hours))],
```
Comment (iandees) [inline locations/spiders/farmershomefurniture.py:52]: This array access construct seems more complicated than it needs to be. The days are "Mon", "Tue", etc. You could just set the `day` argument of `add_range` to `day[:2]`

File: locations/spiders/farmershomefurniture.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+                yield scrapy.Request(store_url, callback=self.parse_store, cb_kwargs=properties)
+
+
+    def parse_store(self, response, city, state, address, phone):
+        opening_hours = OpeningHours()
+        store_hours = response.xpath('//div[@class="workspacearea"]/div/div/p/text()').extract()[2:]
+
+        for hours in store_hours:
+            day = hours.strip().split(':')[0]
```
Comment (iandees) [inline locations/spiders/farmershomefurniture.py:48]: Reuse the `hours.strip().split(':')` so you don't have to call it here and two lines down.

**Comment by ashikjm** [comment]
Thanks for the review. I have implemented the changes as per the suggestions.

**Comment by iandees** [comment]
They seem to be blocking the scraper, but I'll still merge it in so we can improve it later, maybe.

**Comment by ashikjm** [comment]
I can work on improving the spider. I run `pipenv run scrapy crawl farmershomefurniture --output-format=geojson --output=output.geojson` from my local dev env and I only get 200 and 404s. 

Could you let me know where you see them blocking so that I work on it.

**Comment by iandees** [comment]
Click the "Checks" tab on the top of this page, then expand the "Run spiders that changed" section and look for the log file URL. For the most recent commit, it's https://placescraper-results.s3.amazonaws.com/ci/331595670/farmershomefurniture/log.txt. If you look at the log there you can see it made a single request that was 403'd.

**Comment by ashikjm** [comment]
Got it. Let me check and will raise a PR when I find something.

**Pull request merged**